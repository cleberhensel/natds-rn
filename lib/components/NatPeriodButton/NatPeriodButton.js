import React from 'react';
import { Platform, TouchableNativeFeedback } from 'react-native';
import PropTypes from 'prop-types';

import {
  CustomView, Title, SubTitle, CheckedIcon, CustomTouchableOpacity,
} from './NatPeriodButton.styles';

export const NatPeriodButton = ({
  selected, title, subtitle, onSelect, iconSrc,
}) => {
  if (!onSelect) {
    throw new Error('Period Button needs an onSelect property to be passed to the component.');
  }

  return (Platform.OS === 'android' ? (
    <TouchableNativeFeedback onPress={onSelect}>
      <CustomView selected={selected}>
        {selected && (
        <CheckedIcon
          source={{ uri: iconSrc }}
          // source temporario
        />
        )}
        {title && (
        <Title selected={selected}>{title}</Title>
        )}
        {subtitle && (
        <SubTitle selected={selected}>{subtitle}</SubTitle>
        )}
      </CustomView>
    </TouchableNativeFeedback>
  ) : (
    <CustomTouchableOpacity onPress={onSelect}>
      <CustomView selected={selected}>
        {selected && (
        <CheckedIcon
          source={{ uri: iconSrc }}
          // source temporario
        />
        )}
        {title && (
        <Title selected={selected}>{title}</Title>
        )}
        {subtitle && (
        <SubTitle selected={selected}>{subtitle}</SubTitle>
        )}
      </CustomView>
    </CustomTouchableOpacity>
  ));
};

NatPeriodButton.propTypes = {
  selected: PropTypes.bool,
  title: PropTypes.string,
  subtitle: PropTypes.string,
  onSelect: PropTypes.func.isRequired,
  iconSrc: PropTypes.string.isRequired,
};
NatPeriodButton.defaultProps = {
  title: null,
  subtitle: null,
  selected: false,
};
