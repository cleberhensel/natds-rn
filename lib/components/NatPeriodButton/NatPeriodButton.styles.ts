import styled from "styled-components/native";
import { colors } from "../../assets/theme/colors/colors";
const contentDefault = `
    align-items: center;
    justify-content: center;
    border: 2px ${colors.primary.yellow};
    border-radius: 6px;
    width: 30%;
    height: 64px;
    padding: 5px;
    margin-horizontal: 5px;
`;
export const CustomView = styled.View`
    ${contentDefault}
    background-color: ${props =>
      props.selected
        ? colors.content.selectedBackground
        : colors.primary.white};
`;
export const Title = styled.Text`
  font-weight: bold;
  color: ${props => (props.selected ? colors.bw.black : colors.bw.grayDarker)};
`;
export const SubTitle = styled.Text`
  color: ${props => (props.selected ? colors.bw.black : colors.bw.grayDark)};
`;
export const CheckedIcon = styled.Image`
  width: 18;
  height: 18;
  position: absolute;
  top: 1;
  right: 1;
  align-self: flex-end;
`;
export const CustomTouchableOpacity = styled.TouchableOpacity`
  border-radius: 6px;
`;
