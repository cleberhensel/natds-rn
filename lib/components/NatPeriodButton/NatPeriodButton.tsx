/* eslint-disable complexity */
import React from 'react';
import { Platform, TouchableNativeFeedback } from 'react-native';
import {
  CustomView,
  Title,
  SubTitle,
  CheckedIcon,
  CustomTouchableOpacity,
} from './NatPeriodButton.styles';

type NatPeriodButtonProps = {
  iconSrc: string,
  onSelect: (...args: any[]) => any,
  selected?: boolean,
  subtitle?: string,
  title?: string
};
export const NatPeriodButton: React.SFC<NatPeriodButtonProps> = ({
  selected,
  title,
  subtitle,
  onSelect,
  iconSrc,
}) => {
  if (!onSelect) {
    throw new Error(
      'Period Button needs an onSelect property to be passed to the component.',
    );
  }

  return Platform.OS === 'android' ? (
    <TouchableNativeFeedback onPress={onSelect}>
      <CustomView selected={selected}>
        {selected && <CheckedIcon source={{ uri: iconSrc }} />}
        {title && <Title selected={selected}>{title}</Title>}
        {subtitle && <SubTitle selected={selected}>{subtitle}</SubTitle>}
      </CustomView>
    </TouchableNativeFeedback>
  ) : (
    <CustomTouchableOpacity onPress={onSelect}>
      <CustomView selected={selected}>
        {selected && <CheckedIcon source={{ uri: iconSrc }} />}
        {title && <Title selected={selected}>{title}</Title>}
        {subtitle && <SubTitle selected={selected}>{subtitle}</SubTitle>}
      </CustomView>
    </CustomTouchableOpacity>
  );
};
NatPeriodButton.defaultProps = {
  selected: false,
  subtitle: null,
  title: null,
};
