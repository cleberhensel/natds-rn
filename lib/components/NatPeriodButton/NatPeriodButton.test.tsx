import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import { TouchableNativeFeedback, Platform } from 'react-native';
import { NatPeriodButton } from './NatPeriodButton';
import {
  Title,
  CheckedIcon,
  CustomTouchableOpacity,
} from './NatPeriodButton.styles';

const setPlatform = (platform) => {
  Object.defineProperty(Platform, 'OS', {
    get: jest.fn(() => platform),
  });
};

describe('NatPeriodButton', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('Should render the component correctly', () => {
    const PeriodButton = renderer
      .create(<NatPeriodButton onSelect={jest.fn()} />)
      .toJSON();

    expect(PeriodButton).toMatchSnapshot();
  });
  it('Should simulate the select click on the component on the Android platform', () => {
    setPlatform('android');
    const onSelectEvent = jest.fn();
    const wrapper = shallow(<NatPeriodButton onSelect={onSelectEvent} />);

    wrapper
      .find(TouchableNativeFeedback)
      .getElement()
      .props.onPress();
    expect(onSelectEvent.mock.calls.length).toBe(1);
  });
  it('Should simulate the select click on the component on the iOS platform', () => {
    setPlatform('ios');
    const onSelectEvent = jest.fn();
    const wrapper = shallow(<NatPeriodButton onSelect={onSelectEvent} />);

    wrapper
      .find(CustomTouchableOpacity)
      .getElement()
      .props.onPress();
    expect(onSelectEvent.mock.calls.length).toBe(1);
  });
  it('Should the correct styling properties when selected', () => {
    const wrapper = shallow(<NatPeriodButton selected onSelect={jest.fn()} />);

    expect(wrapper.find(CheckedIcon).exists()).toBeTruthy();
  });
  it('Should render header component with title', () => {
    const wrapper = shallow(
      <NatPeriodButton title="Title test" onSelect={jest.fn()} />,
    );

    expect(wrapper.find(Title).text()).toBe('Title test');
  });
  it('Should throw exception when the onSelect property is not passed to the component', () => {
    // @ts-ignore
    expect(() => shallow(<NatPeriodButton />)).toThrow(
      'Period Button needs an onSelect property to be passed to the component.',
    );
  });
});
