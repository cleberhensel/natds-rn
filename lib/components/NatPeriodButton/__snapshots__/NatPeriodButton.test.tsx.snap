// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NatPeriodButton Should render the component correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderRadius": 6,
      "opacity": 1,
    }
  }
>
  <View
    selected={false}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderColor": "#F4AB34",
          "borderRadius": 6,
          "borderStyle": "solid",
          "borderWidth": 2,
          "height": 64,
          "justifyContent": "center",
          "marginHorizontal": 5,
          "paddingBottom": 5,
          "paddingLeft": 5,
          "paddingRight": 5,
          "paddingTop": 5,
          "width": "30%",
        },
      ]
    }
  />
</View>
`;
